Adds Generalized IOU

oke-aditya opened the pull request:
Fixes #2545 

As discussed there, Generalized IOU would be a good feature to add. 
I have added that.
The reference implementation is from detr as suggested by fmassa.

Detr has a slightly different implementation of `box_iou` from torchvision which is used by `generalized_iou`. 
We return `union` and `iou` both in detr, whereas we return only `iou` in torchvision.

I thought not to change the API and instead hence did NOT re-use `box_iou` in `generalized_iou`. 
This ensures both these functions would return a single tensor parameter.

I did a bit of code duplication which might not be best practice, please let me know if that can be avoided. 
I had thought of making a helper function that returns both `iou` and `union` from which we choose in both cases. 

- [x] Adds Generalized IoU implementation.
- [x] Add Tests.
- [x] Add docs.

For adding tests, I could not find any class in `test_ops.py` file for IOU. I could see for NMS etc, I'm unsure where IoU tests are located.

Maybe I overlooked or looked at the wrong place. Do let me know where and how the tests should be!

For docs I saw torchvision [ops.rst ](https://github.com/pytorch/vision/blob/master/docs/source/ops.rst) file. But I cannot find docs for IoU. I'm unsure where to add docs.

Let me know where are they needed.

cc: @fmassa @pmeier

pmeier commented:
@oke-aditya I'm unsubscribing for now. Ping me again when this is ready for review, i.e. not a draft anymore.

oke-aditya commented:
Docs are not added for some of the `boxes.py` codebase in [ops.rst](https://github.com/pytorch/vision/blob/master/docs/source
/ops.rst) file.

I added docs for `iou` and as well as `generalized_iou`. Since they both use `box_area` I added them too. 

Tested these docs locally.

oke-aditya commented:
For tests, I thought to keep it similar to IoU tests. But I cannot find tests for IoU and I am unsure how and where to add it.

pmeier commented:
Hey @oke-aditya, I'm pretty covered right now, so reviewing this will take a while. Sorry for the delay.

oke-aditya commented:
No issues @pmeier 😀

oke-aditya commented:
I had to fix this PR due to Merge conflict of my previous PR. Linter failed here.
``` 
./torchvision/ops/__init__.py:18:49: E231 missing whitespace after ','
```

fmassa reviewed (commented):
Hi @oke-aditya 

Thanks for the PR and sorry for the delay in reviewing!

The way you did it (without re-using `box_iou`) seems ok to me.
For the docs I think everything is good now!

For tests, I propose that we cover a couple of known cases, like Generalized IoU between boxes which are the same, boxes that do not overlap at all, etc (like, at least the sign of the boxes should be known).

We can also take the result from the DETR implementation for a couple of pre-defined boxes and compare against our implementation here.

oke-aditya commented:
I added tests for `box_iou` `generalized_iou` and `box_area`. I covered some cases where IoU is 0, 1 and 0.25. 

For giou they are inf, 1 and 0.25. Please let me know if they right.

fmassa commented on test/test_ops.py:
```diff
@@ -647,5 +647,50 @@ def test_convert_boxes_to_roi_format(self):
+
+        # Expected IoU matrix for these boxes
+        # Expect infinity since intersection is 0 and we divide by it
+        expected = torch.tensor([[1.0, 3.25, float("Inf")], [3.25, 1.0, float("Inf")],
+                                [float("Inf"), float("Inf"), 1.0]])
```
Hum, this looks weird -- Generalized IoU is supposed to be between -1 and 1, where negative values correspond to boxes that do not overlap.

Indeed, taking the implementation from DETR and passing it through `boxes1` and `boxes2`, I get
```python
In [2]: boxes1 = torch.tensor([[0, 0, 100, 100], [0, 0, 50, 50], [200, 200, 300, 300]], dtype=torch.float)

In [3]: boxes2 = torch.tensor([[0, 0, 100, 100], [0, 0, 50, 50], [200, 200, 300, 300]], dtype=torch.float)

In [4]: r =box_ops.generalized_box_iou(boxes1, boxes2)

In [5]: r
Out[5]:
tensor([[ 1.0000,  0.2500, -0.7778],
        [ 0.2500,  1.0000, -0.8611],
        [-0.7778, -0.8611,  1.0000]])
```

oke-aditya replied:
Tests are updated 👍

fmassa commented on torchvision/ops/boxes.py:
```diff
@@ -175,3 +174,40 @@ def box_iou(boxes1: Tensor, boxes2: Tensor) -> Tensor:
+    union = area1[:, None] + area2 - inter
+
+    iou = inter / union
+
```
In here you are missing the remaining of the GIoU computation, which computes the boxes smallest enclosing bounding box, which is [implemented in DETR in this part](https://github.com/facebookresearch/detr/blob/699bf53f3e3ecd4f000007b8473eda6a08a8bed6/util/box_ops.py#L55-L59).

I think the thing that is missing is that
```python
lt = torch.min(boxes1[:, None, :2], boxes2[:, :2])
rb = torch.max(boxes1[:, None, 2:], boxes2[:, 2:])
```
is not the same as what we compute in `box_iou` (note the inverted `min` / `max`). So you need to recompute the `wh` and `area` for those new boxes.

oke-aditya replied:
Absolutely! I missed the gIoU implementation code a bit. I guess I did calculate IoU and union of boxes but missed the next computation (assuming that it was same).

The min and max are inverted which I completely mistook and thought were double calculations and hence did the last step directly :(

I guess I need to add all these 4 lines.

```
    lti = torch.min(boxes1[:, None, :2], boxes2[:, :2])
    rbi = torch.max(boxes1[:, None, 2:], boxes2[:, 2:])
    whi = (rbi - lti).clamp(min=0)  # [N,M,2]
    areai = wh[:, :, 0] * wh[:, :, 1]
    return iou - (areai - union) / areai
```

oke-aditya replied:
Fixes this working fine

fmassa reviewed (changes requested):
Thanks for the tests, they look great!

I think there is a bug with the implementation, let me know what you think

oke-aditya commented:
Ok, so there is small torch issue why the tests are failing. I updated and seem to get correct result as above.

```
boxes1 = torch.tensor([[0, 0, 100, 100], [0, 0, 50, 50], [200, 200, 300, 300]], dtype=torch.float)
boxes2 = torch.tensor([[0, 0, 100, 100], [0, 0, 50, 50], [200, 200, 300, 300]], dtype=torch.float)
out = generalized_box_iou(boxes1, boxes2)
expected = torch.tensor([[1.0, 0.25, -0.7778], [0.25, 1.0, -0.8611], [-0.7778, -0.8611, 1.0]], dtype=torch.float)
```

The issue is out is
```
tensor([[ 1.0000,  0.2500, -0.7778],                                                                                                                                                                                       [ 0.2500,  1.0000, -0.8611],                                                                                                                                                                                       [-0.7778, -0.8611,  1.0000]]) 
```
where `-0.7778` is rounded off. Whereas `-0.7778` is not rounded off in `expected` hence their comparison is false!

Is there any way to compare these two or declare such a float variable `-0.7778888888..` variable. Or some way to assert them?

Strange! implementation is fine now, but tests aren't :(

fmassa commented:
One way to do it is to use `(tensor1 - tensor2).abs().max() < tolerance`. We use this a lot in the codebase, specially when dealing with floating point data. I would use something like that as well here

oke-aditya commented:
Excellent advice. It's a new thing I learnt and will remember for next time too.
I guess now this PR is ready to merge. 👍

fmassa reviewed (approved):
Thanks a lot!
